### PR TITLE
Merging the the staged fork into develop

### PR DESCRIPTION
**Pull request by maasva**
### Description
This is a big one. This PR includes the following: 

1. Refactored ff_prep to adhere to SE principles
2. Test case suite for ff_prep, ff_structurecheck, ff_run
3. Environment variables can be set though config files, think of data folder and such
4. Added testing data for BRN
5. Set up CI pipelines that do building, testing and styling. 

### Requirements
This PR fulfills part/gives examples for the Project goal of improving the codebase quality and getting it ready for open-source contributions. 

### Configurations

NA

### Related Issues

#6

### CODE REVIEW

File: .github/workflows/styler_lintr.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
+name: Improving Readability
+on:
+  pull_request:
+    branches: 
+      - "develop"
+      - "main"
```
Comment (aminbksi) [inline .github/workflows/styler_lintr.yml:6]: please remove the check lintr and styler on main

File: .github/workflows/styler_lintr.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+          echo "changes_made=false" >> $GITHUB_ENV
+        else
+          echo "Changes made by styler."
+          echo "changes_made=true" >> $GITHUB_ENV
+        fi
+
+    - name: Commit and push if styler made a change
+      if: env.changes_made == 'true'
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (aminbksi) [inline .github/workflows/styler_lintr.yml:41]: @hasruleng I just noticed, how come we have github token here an not for the tests?

File: R/ff_analyze.R [hunk 3]
```diff
@@ -25,72 +25,106 @@
...
-                       date=NULL, tile=NULL, method=NA, addWKT = FALSE, verbose=FALSE){
-  if (!(class(predictions) %in% c("character","SpatRaster"))) {stop("predictions is not a raster or path to a raster")}
-  if (!(class(groundtruth) %in% c("character","SpatRaster"))) {stop("predictions is not a raster or path to a raster")}
-  if (!is.null(csvfile)) {if (append == TRUE & !file.exists(csvfile)) {
-    append <- FALSE
-    ff_cat("CSV file did not yet exist, creating empty one", verbose = verbose)
-    }}
+ff_analyze <- function(predictions, groundtruth, forestmask = NULL, csvfile = NULL, country = NULL,
+                       append = TRUE, analysis_polygons = NULL, return_polygons = TRUE, remove_empty = TRUE,
+                       date = NULL, tile = NULL, method = NA, addWKT = FALSE, verbose = FALSE) {
```
Comment (aminbksi) [inline R/ff_analyze.R:30]: what is addWKT?

File: R/ff_analyze.R [hunk 4]
```diff
@@ -25,72 +25,106 @@
...
       pastdata$X <- NULL
-      write.csv(rbind(pastdata,pol_df),csvfile)}else{
-        if (!file.exists(csvfile) & append & verbose) {ff_cat("the given file does not exist, while append was set to TRUE",color = "yellow", verbose = verbose)}
-        write.csv(pol_df,csvfile)
+      write.csv(rbind(pastdata, pol_df), csvfile)
+    } else {
+      if (!file.exists(csvfile) & append & verbose) {
+        ff_cat("the given file does not exist, while append was set to TRUE", color = "yellow", verbose = verbose)
       }
+      write.csv(pol_df, csvfile)
```
Comment (aminbksi) [inline R/ff_analyze.R:124]: can we have a more clear variable for pol_df

File: R/filter_by_feature.R [hunk 5]
```diff
@@ -2,69 +2,65 @@
...
 #' @param matrix The input matrix to filter.
 #' @param verbose Logical indicating whether to display verbose output (default is TRUE).
 #'
 #' @return A list containing the filtered matrix and the indices of the filtered rows.
 #'
 #' @examples
 #' filter_by_feature(
-#'   fltr_features = c("landpercentage", "forestmask"),
-#'   fltr_condition = c(">100", ">0"),
+#'   filter_features = c("landpercentage", "forestmask"),
```
Comment (aminbksi) [inline R/filter_by_feature.R:14]: not sure but if these are happen to change, it would be better to have them in the config file

**Review by aminbksi — CHANGES_REQUESTED** [review]
very impressive ff_prep, but also made some comments

**Review by jduijvenbode-WWF — APPROVED** [review]
I can't find any flaws!

File: R/ff_analyze.R (hunk 3 above)
Comment (jduijvenbode-WWF) [reply R/ff_analyze.R:30]: addWKT is a method to add the coordinates and class of a spatial dataset as an attribute. it's for vector data (polygons, lines and points). it is used in powerbi

**Pull request merged**